### ISSUE TITLE
Docs of input/output annotations should state that usage on Kotlin properties is supported

### ISSUE THREAD

**Issue by marcphilipp** [gradle/gradle#11927]
Declaring the following Kotlin property in a `Task` implementation, causes the value of `myInput` to be tracked as an input.

```kotlin
@Input
val myInput = objects.property<String>()
```

### Expected Behavior

The Javadoc of `@Input` currently states:

> This annotation should be attached to the getter method in Java or the property in Groovy.

### Current Behavior

It should state:

> This annotation should be attached to the getter method in Java or the property in Groovy or Kotlin.

### Context

Our Kotlin samples all use `@get:Input` so it's unclear if this is officially supported.

Related issue: #9030

**Comment by wolfs** [comment]
+1 to adding something to the Javadoc about Kotlin.

I think `@get:Input` is better, since then only the getter is annotated, and that is what we look out for. So I would use that in samples and in Javadoc. For Groovy, there is no shorthand to only annotate the getter.

**Comment by lptr** [comment]
I don't agree about using the `@get:` prefix. What we want is code that is easily readable/reasoned about. It is the property that represents an input; that it has a getter is irrelevant. So I'd actually encourage people to use `@Input` in Kotlin.

**Comment by marcphilipp** [comment]
I also find using just `@Input` more readable and logical.

**Comment by big-guy** [comment]
We should update everything to use managed properties.  Instead of:
```
@Input
val myInput = objects.property<String>()
```

It should be:
```
@Input
abstract fun getMyInput(): Property<String>
```

That would make it the same everywhere and avoid annotations on Kotlin properties.

**Comment by marcphilipp** [comment]
To me that doesn't feel like idiomatic Kotlin.

FWIW our [sample on managed properties](https://docs.gradle.org/current/userguide/custom_gradle_types.html#read_only_managed_properties) does this:

```kotlin
@get:Input
abstract val uri: Property<URI>
```

**Comment by big-guy** [comment]
Even better, +1 for dropping `get:`, so we'd use:

```
@Input
abstract val uri: Property<URI>
```

There's no setter, so restricting the annotation doesn't matter/make sense.

**Comment by big-guy** [comment]
I also think we should just update the docs exactly as Marc suggested.  It's correct and if it's less confusing for now, that's a net positive.

**Comment by wolfs** [comment]
We should improve the Javadoc.

**Comment by chkpnt** [comment]
Mh, reading this thread I thought I can drop `get:`, but this doesn't work: I get `This annotation is not applicable to target 'member property without backing field or delegate'`

```kotlin
@InputDirectory
abstract val inputDir: Property<Path>
```

**Comment by fab1an** [comment]
Hi. from this issue it is not clear to me whether just specifying the annotation on a property (without `get:`) works and the java-doc is wrong, or it doesn't work.

My opinion is it should work.
